x-pack/filebeat/input/awss3/parser.go

```diff
@@ -0,0 +1,77 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/cfgtype"
+	"github.com/elastic/beats/v7/libbeat/reader"
+	"github.com/elastic/beats/v7/libbeat/reader/multiline"
+	"github.com/elastic/beats/v7/libbeat/reader/readfile"
+)
+
+var (
+	ErrNoSuchParser = errors.New("no such parser")
+)
+
+// parser transforms or translates the Content attribute of a Message.
+// They are able to aggregate two or more Messages into a single one.
+type parser interface {
+	io.Closer
+	Next() (reader.Message, error)
+}
+
+type parserConfig struct {
+	maxBytes       cfgtype.ByteSize
+	lineTerminator readfile.LineTerminator
+}
+
+func newParsers(in reader.Reader, pCfg parserConfig, c []common.ConfigNamespace) (parser, error) {
```

Comment:
@kvch this looks like we have to duplicate parser setup from the filestream input here. This is somewhat unfortunate, especially if we want to add similar functionality to other inputs. How about providing a common package for parser setup?

Inputs that want to make use of parsers should be able to do it more or less like this:


```

type Config struct {
  ...

  Parser inpparser.Settings
  ...
}

...

parser, err := config.Parser.Create(reader)
...

```